src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -133,174 +156,99 @@
-      (update! (global-metric :retry-counts) retries)
-      (update! (cmd-metric command version :retry-counts) retries)
-
-      (when (>= retries max-retries)
-        (mark! (global-metric :discarded))
-        (mark! (cmd-metric command version :discarded))
-        (on-discard msg))
-
-      (when (< retries max-retries)
-        (let [result (time! (cmd-metric command version :processing-time)
-                            (f msg))]
-          (mark! (cmd-metric command version :processed))
-          result)))))
-
-(defn wrap-with-thread-name
-  "Wrap a message processor `f` such that the calling thread's name is
-  set to `<prefix>-<thread-id>`. This is useful for situations where
-  thread names are ordinarily duplicated across multiple threads, or
-  if the default names aren't descriptive enough."
-  [f prefix]
-  (fn [msg]
-    (let [t    (Thread/currentThread)
-          name (format "%s-%d" prefix (.getId t))]
-      (when-not (= (.getName t) name)
-        (.setName t name))
-      (f msg))))
-
-(defn handle-command-discard
-  [{:keys [command annotations] :as msg} discard]
-  (let [attempts (count (:attempts annotations))
-        id       (:id annotations)
-        certname (get-in msg [:payload :certname])]
-    (log/error (i18n/trs "[{0}] [{1}] Exceeded max {2} attempts for {3}" id command attempts certname))
-    (discard msg nil)))
-
-(defn handle-parse-error
-  [msg e discard]
-  (log/error e (i18n/trs "Fatal error parsing command: {0}" msg))
-  (discard msg e))
-
-(defn handle-command-failure
-  "Dump the error encountered during command-handling to the log and discard
-  the message."
-  [{:keys [command annotations payload] :as msg} e discard]
-  (let [attempt (count (:attempts annotations))
-        id (:id annotations)
-        certname (:certname payload)]
-    (log/error e (i18n/trs "[{0}] [{1}] Fatal error on attempt {2} for {3}"
-                           id command attempt certname))
-    (-> msg
-        (annotate-with-attempt e)
-        (discard e))))
-
 ;; The number of times a message can be retried before we discard it
 (def maximum-allowable-retries 5)
 
-(defn handle-command-retry
-  "Dump the error encountered to the log, and re-publish the message,
-  with an incremented retry counter, after a ten minute delay."
-  [{:keys [command version annotations payload] :as msg} e publish-fn]
-  (mark! (cmd-metric command version :retried))
-  (let [attempt (count (:attempts annotations))
-        id (:id annotations)
-        certname (:certname payload)
-        delay (mq/delay-property 10 :minutes)]
-    (log/error e (i18n/trs "[{0}] [{1}] Retrying after attempt {2} for {3}, due to: {4}"
-                           id command attempt certname e))
-    (time! (global-metric :retry-persistence-time)
-           (-> msg
-               (annotate-with-attempt e)
-               json/generate-string
-               (publish-fn delay)))))
-
-(defn create-message-handler
```

Comment:
I've collapsed down most of this middleware. I think there is some simplification that can be done beyond this, but my initial goal was being able to see all of the pieces together. I think with stockpile, this code is much easier to test and breaking it up the way it was before is no longer necessary.